From e760681d069112dd7af570d6820a33273a5729f6 Mon Sep 17 00:00:00 2001
From: Horia Geanta <horia.geanta@freescale.com>
Date: Tue, 20 May 2014 14:44:52 +0300
Subject: [PATCH 1045/1089] fsl_qbman: fix access to alloc->used list in DPA
 allocator

[Original patch taken from QorIQ-SDK-V1.6-SOURCE-20140619-yocto.iso]

The operation of adding the newly allocated node in alloc->used list
must be protected by the alloc->lock.

Change-Id: I8f80006b59102dd4563f36d0d196afe3c5307489
Signed-off-by: Horia Geanta <horia.geanta@freescale.com>
Tested-by: Mircea Pop <mircea.pop@freescale.com>
Reviewed-on: http://git.am.freescale.net:8181/12525
Reviewed-by: Alexandru Porosanu <alexandru.porosanu@freescale.com>
Tested-by: Review Code-CDREVIEW <CDREVIEW@freescale.com>
Reviewed-by: Bogdan Hamciuc <bogdan.hamciuc@freescale.com>
Reviewed-by: Jose Rivera <German.Rivera@freescale.com>
---
 drivers/staging/fsl_qbman/dpa_alloc.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/fsl_qbman/dpa_alloc.c b/drivers/staging/fsl_qbman/dpa_alloc.c
index 545adc9..2468a97 100644
--- a/drivers/staging/fsl_qbman/dpa_alloc.c
+++ b/drivers/staging/fsl_qbman/dpa_alloc.c
@@ -498,8 +498,10 @@ done:
 		list_del(&i->list);
 		kfree(i);
 		*result = base;
+	} else {
+		spin_unlock_irq(&alloc->lock);
 	}
-	spin_unlock_irq(&alloc->lock);
+
 err:
 	DPRINT("returning %d\n", i ? num : -ENOMEM);
 	DUMP(alloc);
@@ -508,13 +510,16 @@ err:
 
 	/* Add the allocation to the used list with a refcount of 1 */
 	used_node = kmalloc(sizeof(*used_node), GFP_KERNEL);
-	if (!used_node)
+	if (!used_node) {
+		spin_unlock_irq(&alloc->lock);
 		return -ENOMEM;
+	}
 	used_node->base = *result;
 	used_node->num = num;
 	used_node->refcount = 1;
 	used_node->is_alloced = 1;
 	list_add_tail(&used_node->list, &alloc->used);
+	spin_unlock_irq(&alloc->lock);
 	return (int)num;
 }
 
-- 
2.0.2

